daphne_worker/src/durable/aggregate_store.rs

```diff
@@ -47,6 +53,137 @@ pub struct AggregateStore {
+                daphne::vdaf::VdafAggregateShare::Field64(_) => VdafKind::Field64,
+                daphne::vdaf::VdafAggregateShare::Field128(_) => VdafKind::Field128,
+                daphne::vdaf::VdafAggregateShare::FieldPrio2(_) => VdafKind::FieldPrio2,
+            }),
+        };
+
+        (this, share.data)
+    }
+
+    fn into_agg_share_with_data(self, data: daphne::vdaf::VdafAggregateShare) -> DapAggregateShare {
+        DapAggregateShare {
+            data: Some(data),
+            ..self.into_agg_share_without_data()
+        }
+    }
+
+    fn into_agg_share_without_data(self) -> DapAggregateShare {
+        DapAggregateShare {
+            report_count: self.report_count,
+            min_time: self.min_time,
+            max_time: self.max_time,
+            checksum: self.checksum,
+            data: None,
+        }
+    }
+}
+
+fn js_map_to_chunks<T: DeserializeOwned>(keys: &[String], map: js_sys::Map) -> Vec<T> {
+    keys.iter()
+        .map(|k| JsValue::from_str(k))
+        .filter(|k| map.has(k))
+        .map(|k| map.get(&k))
+        .flat_map(|js_v| {
+            serde_wasm_bindgen::from_value::<Vec<T>>(js_v).expect("expect an array of bytes")
+        })
+        .collect()
+}
+
+impl AggregateStore {
+    fn agg_share_shard_keys(&self) -> Vec<String> {
+        (0..MAX_CHUNK_KEY_COUNT)
+            .map(|n| format!("chunk_v2_{n:03}"))
+            .collect()
+    }
+
+    async fn get_agg_share(&self, keys: &[String]) -> Result<DapAggregateShare> {
+        let all_keys = keys
+            .iter()
+            .map(String::as_str)
+            .chain([METADATA_KEY])
+            .collect::<Vec<_>>();
+        let values = self.state.storage().get_multiple(all_keys).await?;
+
+        if values.size() == 0 {
+            return Ok(DapAggregateShare::default());
+        }
+
+        let meta_key = JsValue::from_str("meta");
+        let meta =
+            serde_wasm_bindgen::from_value::<DapAggregateShareMetadata>(values.get(&meta_key))
+                .unwrap_or_else(|e| {
+                    tracing::error!("failed to deser DapAggregateShareMeta: {e:?}");
+                    panic!("{e}")
+                });
+
+        let chunks = js_map_to_chunks(keys, values);
+
+        Ok(if chunks.is_empty() {
+            meta.into_agg_share_without_data()
+        } else {
+            let kind = meta.kind.expect("if there is data there should be a type");
+
+            fn from_slice<T: FieldElement>(chunks: &[u8]) -> Result<AggregateShare<T>> {
+                let share = T::byte_slice_into_vec(chunks).map_err(|e| {
+                    worker::Error::Internal(
+                        serde_wasm_bindgen::to_value(&e.to_string())
+                            .expect("string never fails to convert to JsValue"),
+                    )
+                })?;
+                // TODO: this an abuse of this API, this type should not be constructed this way.
```

Comment:
Please file an issue as well. We should make sure libprio-rs supports our use case here.
```suggestion
                // TODO(mendess): this an abuse of this API, this type should not be constructed this way.
```